### PR TITLE
adding bias correlation

### PR DESCRIPTION
**Pull request by ryanharvey1**
Attempt to add bias correlation

### CODE REVIEW

**Comment by kushaangupta** [comment]
@ryanharvey1 we need to re-evaluate the non-significance test of `PairwiseBias`, it works for some iterations but not all. I don't understand the logic behind it and why should we expect it to be non-significant always

**Comment by ryanharvey1** [comment]
In the all significant case, we explicitly define sequential firing.

```
    n_spikes = 200
    spk_1 = np.sort(np.random.rand(n_spikes))
    post_spikes = []
    post_neurons = []
    for i_neurons in range(n_neurons):
        post_spikes.append(spk_1 + i_neurons * 0.01)
        post_neurons.append(np.ones(n_spikes) * i_neurons)
```

For the non-significant, the idea is that we are adding random offsets for each neuron, so there should not be greater sequential firing than chance. This case would be better if we do  `np.sort(np.random.rand(n_spikes)` instead of `spk_1 + np.sort(np.random.rand(n_spikes)` so that each neuron has no systematic offset from each other. 
```
    # test non significant intervals

    n_spikes = 200
    spk_1 = np.sort(np.random.rand(n_spikes))
    post_spikes = []
    post_neurons = []
    for i_neurons in range(n_neurons):
        post_spikes.append(spk_1 + np.sort(np.random.rand(n_spikes)))
        post_neurons.append(np.ones(n_spikes) * i_neurons)

    post_spikes = np.concatenate(post_spikes)
    post_neurons = np.concatenate(post_neurons)

    z_score, p_value, cosine_val = transformer.transform(
        post_spikes, post_neurons, post_intervals
    )
    print(f"Z-score: {z_score}, P-value: {p_value}")

    assert z_score.size == post_intervals.shape[0]
    assert p_value.size == post_intervals.shape[0]
    assert (p_value > 0.05).all()
```

File: neuro_py/ensemble/pairwise_bias_correlation.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+                    nspikes_ij = np.searchsorted(
+                        spikes_i, spikes_j, side='right').sum()
+                    bias[i, j] = nspikes_ij / (nspikes_i * nspikes_j)
+                else:
+                    bias[i, j] = fillneutral  # Neutral bias if no spikes
+
+    return bias
+
+
+def bias_matrix_fast(
```
Comment (ryanharvey1) [inline neuro_py/ensemble/pairwise_bias_correlation.py:78]: do we need both normal and fast version? What is the trade off for fast? If nothing, we can just keep the fast

File: neuro_py/ensemble/pairwise_bias_correlation.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    x = np.nan_to_num(x)
+    y = np.nan_to_num(y)
+
+    cossim = sklearn.metrics.pairwise.cosine_similarity(x, y)
+
+    # Compute cosine similarity
+    return cossim.item()
+
+
+def observed_and_shuffled_correlation(
```
Comment (ryanharvey1) [inline neuro_py/ensemble/pairwise_bias_correlation.py:208]: add static typing and docstring

File: neuro_py/ensemble/replay.py [hunk 3]
```diff
@@ -398,3 +367,591 @@ def trajectory_score_bst(
             np.array(weighted_corr_col_cycle).T,
         )
     return scores, weighted_corr
+
```
Comment (ryanharvey1) [inline neuro_py/ensemble/replay.py:370]: we could delete all this commented out code if we don't need it

File: neuro_py/ensemble/replay.py [hunk 4]
```diff
@@ -398,3 +367,591 @@ def trajectory_score_bst(
...
+                    post_spikes,
+                    post_neurons,
+                    self.task_normalized,
+                    post_intervals,
+                    interval_i,
+                )
+                for interval_i in range(post_intervals.shape[0])
+            )
+        )
+        # non parallel version
```
Comment (ryanharvey1) [inline neuro_py/ensemble/replay.py:890]: we can add non parallel option for debugging in the future

File: tests/test_replay.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            post_spikes_nonsig, post_neurons_nonsig, post_nonsig_seq_epochs
+        )
+
+        # Check results are of expected shape and values are valid
+        self.assertEqual(len(z_score_sig), len(post_sig_seq_epochs))
+        self.assertEqual(len(z_score_nonsig), len(post_nonsig_seq_epochs))
+
+        # Check that significant replay has lower p-values than non-significant replay
+        self.assertTrue(np.all(p_value_sig < 0.05))
+        self.assertTrue(np.mean(p_value_sig) < np.mean(p_value_nonsig))
```
Comment (ryanharvey1) [inline tests/test_replay.py:125]: I'm still a bit confused why a random sequence would occasionally come up as more sequential than chance levels

Comment (kushaangupta) [reply tests/test_replay.py:125]: just accounting for stochasticity due to small number of shuffles. If we increase num_shuffles that discrepancy goes away but just to be on the safe side this remains a more robust test case

File: neuro_py/ensemble/pairwise_bias_correlation.py (hunk 1 above)
Comment (kushaangupta) [reply neuro_py/ensemble/pairwise_bias_correlation.py:78]: I like the normal version for readability, easy to understand the code. If there's only fast version in the codebase, it can be quite hard to interpret.

**Comment by ryanharvey1** [comment]
you have linting issues still

**Comment by kushaangupta** [comment]
@ryanharvey1 Should be good to review/merge now

**Comment by ryanharvey1** [comment]
looks good!
